RFCs/0014-TestCase-FullyQualifiedName.md

```diff
@@ -0,0 +1,48 @@
+This note clarifies the guidelines that test adapters should consider when setting the FullyQualifiedName on test cases before returning them from the discovery process.
+
+## Overview
+There are a broad variety of choices that current Test Adapter implementers have made when setting the FullyQualifiedName on TestCase objects created during test discovery. This note lays out some basic guidelines to make the usage of FullyQualifiedName more useful across adapters.
+
+## Usage of FullyQualifiedName
+The Test Platform uses FullyQualifiedName as a string moniker to name tests during discovery. Visual Studio may filter tests by FullyQualifiedName in order to more quickly filter and run specific tests. The guidelines below clarify the usage of FullyQualifiedName so that these optimized test execution scenarios can be realized.
+
+## Guidelines
+1. **FullyQualifiedName does NOT need to be unique across your test set.** Internally, TestCase objects are tracked by a unique ID (GUID) which can either be set by the Test Adapter implementor, or automatically generated otherwise. The FullyQualifiedName should be unique if possible, but it's not a requirement.
+1. **FullyQualifiedName should be deterministic** FullyQualifiedName should have the same value for the same test across different discovery sessions. This means that including random data, time-dependent data or environment specific data in the FullyQualifiedName should be avoided.
+1. **FullyQualifiedName is expected to be a reference to the method that implements the test** The Test Platform is expecting the FullyQualifiedName to refer to a particular test case. If there are multiple tests that are instantiated from a test method (data-driven tests for instance), then the FullyQualifiedName should distinguish these using parentheses to represent data parameters and angle brackets to represent type parameters. See below.
+
+## Recommended Form
+FullyQualifiedName is expected to be a reference to the TestCase method that implements the test (or tests). The FullyQualifiedName should be unique to a particular TestCase instance, if possible.
+
+The recommended forms of FullyQualifiedName are described by the following examples:
+
+    Namespace.Namespace.Class.Method
+    Namespace.Namespace.Class.Method()
+    Namespace.Namespace.Class<Int32>.Method()
+    Namespace.Namespace.Class<Int32>.Method<String>()
+    Namespace.Namespace.Class+InnerClass.Method
+    Namespace.Namespace.Class.Method(x: 1, msg: "string")
+    Namespace.Namespace.Class.Method(1, "string")
+    Namespace.Namespace.Class(init: "Value").Method(param: 23)
+    Namespace.Namespace.Class<String>("Value").Method(23)
+
+## Normalized Form
+When referencing test cases by FullyQualifiedName, the test platform recognizes a normalized form of the FullyQualifiedName. This normalized form is the FullyQualifiedName with segments in parentheses and segments in angle brackets removed. The normalized form will not be unique among TestCases.
```

Comment:
I'm guessing we would be adding an extra property of NormalizedFQN on TestCase object. This would result in a json protocol change.

Reply:
I don't think we need to add a property yet, but it may be helpful in the future.

Reply:
Not clear to me in what scenario having a NFQN property on TC would be beneficial. it seems the only reason we are interested in NFQN is so we can have a placeholder TC in test window, which can be used to run all related tests by filter.

Ultimately, we probably need to have another way to run tests by name, other than using TestCaseFilter (we use filter mostly because it does most things we wanted with only minimal change), say by specifying a list of `(string source, string[] fqns)` so we eliminate the ambiguity and have the ability to run just one input of a theory test such as `N.C.M(x: 24)`, plus a chance to further improve perf (for example, we can have a API that simply passing those source-fqns pairs to adapter and let it decide what to do, instead of the callback model currently utilized by filter)

Reply:
Sounds good.

Reply:
Please see comments on NUnit PR https://github.com/nunit/nunit3-vs-adapter/pull/405#issuecomment-340079832 to implement a normalized FQN in the adapter.  This has unwanted sideeffects in the Test Explorer.  The display name must match the FQN.  
Further, the following warning is seen in Windows test output 
[10/27/2017 9:32:03 PM Warning] A test with the same name 'NUnit.VisualStudio.TestAdapter.Tests.TestDiscoveryTests(NUnit.VisualStudio.TestAdapter.Tests.Fakes.FakeDiscoveryContext).VerifyTestCaseIsFound("MockTest3","NUnit.Tests.Assemblies.MockTestFixture.MockTest3")' already exists. This test is not added to the test window.  I assume these are related to the same.

Reply:
@OsirisTerje The display name does not need to be the same as the FQN, but the FQN and the display name should have the same level of uniqueness (ideally completely unique for both). The issue in the referenced PR is that the ID being assigned to the TestCase is not unique (as a consequence of the FQN not being unique, and the Id not being otherwise set).